pkg/server/container_stats_list.go

```diff
@@ -17,14 +17,75 @@ limitations under the License.
 package server
 
 import (
-	"errors"
-
+	"fmt"
 	"golang.org/x/net/context"
 
+	"github.com/containerd/cgroups"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/typeurl"
 	"k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
 )
 
 // ListContainerStats returns stats of all running containers.
 func (c *criContainerdService) ListContainerStats(ctx context.Context, in *runtime.ListContainerStatsRequest) (*runtime.ListContainerStatsResponse, error) {
-	return nil, errors.New("not implemented")
+	request, err := c.buildTaskMetricsRequest(in)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build metrics request: %v", err)
+	}
+	resp, err := c.taskService.Metrics(context.Background(), &request)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch metrics for tasks: %v", err)
+	}
+	criStats, err := toCRIContainerStats(resp.Metrics)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to cri containerd stats format: %v", err)
+	}
+	return criStats, nil
+}
+
+func toCRIContainerStats(stats []*types.Metric) (*runtime.ListContainerStatsResponse, error) {
+	containerStats := new(runtime.ListContainerStatsResponse)
+	for _, stat := range stats {
+		var cs runtime.ContainerStats
+		if err := getContainerMetrics(stat, &cs); err != nil {
+			return nil, fmt.Errorf("failed to decode container metrics: %v", err)
+		}
+		containerStats.Stats = append(containerStats.Stats, &cs)
+	}
+	return containerStats, nil
+}
+
+func getContainerMetrics(stats *types.Metric, cs *runtime.ContainerStats) error {
+	s, err := typeurl.UnmarshalAny(stats.Data)
+	if err != nil {
+		return fmt.Errorf("failed to extract container metrics: %v", err)
+	}
+	metrics := s.(*cgroups.Metrics)
+	cs.Cpu = &runtime.CpuUsage{Timestamp: stats.Timestamp.Unix(), UsageCoreNanoSeconds: &runtime.UInt64Value{Value: metrics.CPU.Usage.Total}}
+	cs.Memory = &runtime.MemoryUsage{Timestamp: stats.Timestamp.Unix(), WorkingSetBytes: &runtime.UInt64Value{metrics.Memory.Usage.Usage}}
+
+	cs.Attributes = &runtime.ContainerAttributes{Id: stats.ID}
+	return nil
+}
+func (c *criContainerdService) buildTaskMetricsRequest(r *runtime.ListContainerStatsRequest) (tasks.MetricsRequest, error) {
+	var req tasks.MetricsRequest
+	if r == nil || r.Filter == nil {
+		return req, nil
+	}
```

Comment:
I believe the relationship between different filters in CRI is `AND`.

What about containerd?
1) If different filters in containerd is `OR`, then it's not correct.
2) If different filters in containerd is `AND`, then we'll get no container, because there is no single container has all the ids.

As is mentioned in our discussion, I'm fine with list all containerstats and do filter afterwards, if the filter doesn't work that well :) Up to you to decide.